openmetadata-ui/src/main/resources/ui/src/hooks/paging/usePaging.ts

```diff
@@ -14,52 +14,123 @@ import {
+
+interface CursorState {
+  cursorType: CursorType | null;
+  cursorValue: string | null;
+}
+
+interface PagingHistoryStateData {
+  cursorData: CursorState;
+  currentPage: number | null;
+  pageSize: number | null;
+}
 
 export interface UsePagingInterface {
   paging: Paging;
   handlePagingChange: Dispatch<SetStateAction<Paging>>;
   currentPage: number;
-  handlePageChange: Dispatch<SetStateAction<number>>;
+  handlePageChange: (
+    page: number | ((page: number) => number),
+    cursorData?: CursorState,
+    pageSize?: number | null
+  ) => void;
   pageSize: number;
   handlePageSizeChange: (page: number) => void;
   showPagination: boolean;
+  pagingCursor: PagingHistoryStateData;
 }
-
 export const usePaging = (
   defaultPageSize = PAGE_SIZE_BASE
 ): UsePagingInterface => {
   const [paging, setPaging] = useState<Paging>(pagingObject);
   const [currentPage, setCurrentPage] = useState<number>(INITIAL_PAGING_VALUE);
   const [pageSize, setPageSize] = useState(defaultPageSize);
+  const history = useHistory();
+  const location = useCustomLocation();
+  const historyState = location.state as any;
+  const pagingCursorHistoryState: PagingHistoryStateData = {
+    cursorData: historyState?.cursorData,
+    currentPage: historyState?.currentPage,
+    pageSize: historyState?.pageSize,
+  };
 
   const handlePageSize = useCallback(
     (page: number) => {
       setPageSize(page);
       setCurrentPage(INITIAL_PAGING_VALUE);
+
+      // Update location state to persist pageSize for navigation
+      history.replace({
+        ...location,
+        state: {
+          pageSize: page,
+        },
+      });
     },
-    [setPageSize, setCurrentPage]
+    [setPageSize, setCurrentPage, location, history]
   );
-
   const paginationVisible = useMemo(() => {
     return paging.total > pageSize || pageSize !== defaultPageSize;
   }, [defaultPageSize, paging, pageSize]);
 
+  const handlePageChange = useCallback(
+    (
+      page: number | ((page: number) => number),
+      cursorData?: CursorState,
+      pageSize?: number | null
+    ) => {
+      setCurrentPage(page);
+      if (cursorData) {
+        history.replace({
+          ...location,
+          state: {
+            ...((location.state as any) || {}),
+            cursorData,
+            currentPage: page,
+            pageSize,
+          },
+        });
+      }
+    },
+    [setCurrentPage, history, location]
+  );
+
+  // set pagesize and current page on page reload from history state if present
+  useEffect(() => {
+    const pageSizeToUse = historyState?.pageSize ?? pageSize;
+    const cursorData = historyState?.cursorData?.cursorType
+      ? historyState.cursorData
+      : { cursorType: null, cursorValue: null };
+    const pageToUse = historyState?.cursorData?.cursorType
+      ? historyState.currentPage
+      : currentPage;
+
+    handlePageSize(pageSizeToUse);
+    handlePageChange(pageToUse, cursorData, pageSizeToUse);
+  }, []);
```

Comment:
This will provide 2 renders instead having it on useEffect, can we pass this as initial values only?

Reply:
addressed, passing it as initial values